### PR TITLE
Add /upgrade_mapblocks command

### PR DESCRIPTION
**Pull request by SmallJoker**
This command is designed to provide the easiest transition possible to update ancient maps.
In the future it will allow to remove support for very old maps (< 0.4.x) to get rid of the legacy code.

Removes unused code, minor code tidy.

## To do

This PR is Ready for Review.

## How to test

1) Have an old map, example: https://forum.minetest.net/viewtopic.php?f=12&t=3364
2) Ensure the log level is set to `action` or lower
3) Change the `world.mt`: `gameid = minetest`
4) Patch `minetest_game/mods/weather/init.lua`:
  * L91 `local humid = minetest.get_humidity(player:get_pos()) or 0`
5) Join the world
6) `/upgrade_mapblocks`
7) Check log output

### CODE REVIEW

File: src/map.cpp [hunk 1]
```diff
@@ -1983,17 +1911,11 @@ void ServerMap::loadBlock(std::string *blob, v3s16 p3d, MapSector *sector, bool
...
-		if(save_after_load)
-			saveBlock(block);
-
-		// We just loaded it from, so it's up-to-date.
-		block->resetModified();
+		// Upgrade ancient mapblock versions on the fly
+		if (version < SER_FMT_VER_LOWEST_WRITE)
+			block->raiseModified(MOD_STATE_WRITE_AT_UNLOAD, MOD_REASON_SER_FMT_UPGRADE);
+		else
+			block->resetModified();
```
Comment (SmallJoker) [inline src/map.cpp:1918]: Sneaky change. This updates mapblocks automatically if they're older than what `MapBlock::serialize` supports. It should spread the load and incrementally update old maps.

Is it acceptable or shall I revert this part?

File: src/map.cpp [hunk 2]
```diff
@@ -1952,7 +1882,7 @@ bool ServerMap::saveBlock(MapBlock *block, MapDatabase *db)
 	return ret;
 }
 
-void ServerMap::loadBlock(std::string *blob, v3s16 p3d, MapSector *sector, bool save_after_load)
```
Comment (SmallJoker) [inline src/map.cpp:1955]: Note to reviewers: Unused argument, was always `false`.

File: src/server.cpp [hunk 3]
```diff
@@ -327,8 +327,10 @@ Server::~Server()
 	}
 
 	// Delete things in the reverse order of creation
-	delete m_emerge;
 	delete m_env;
+	// ServerEnvironment::deactivateFarObjects
+	// requires EmergeManager to be alive
+	delete m_emerge;
```
Comment (SmallJoker) [inline src/server.cpp:333]: Note to reviewers: This fixes a segfault in `deactivateFarObjects` where `EmergeManager` was required in a later stage. Leaving the world now produces warnings, but that's far better than a crash.

File: builtin/game/chat.lua [hunk 4]
```diff
@@ -1039,6 +1039,24 @@ core.register_chatcommand("clearobjects", {
 	end,
 })
 
+core.register_chatcommand("upgrade-mapblocks", {
```
Comment (paramat) [inline builtin/game/chat.lua:1042]: "upgrade_mapblocks" or "upgrademapblocks" for consistency with other chat commands.

**Comment by pyrollo** [comment]
Wouldn't it be better to have some kind of command line utility (or argument) instead of a chat command that will be used only once in a while ?

**Comment by SmallJoker** [comment]
I originally wanted to make it a startup command, but here you need to load the entire map plus node definitions and objects - after all that it results in starting an entire server anyway.

**Comment by pyrollo** [comment]
Then a simple command line argument to allow such conversion. Something like `--migrate` option. So, to upgrade map blocks you would have to run once something like:
```
minetest --world XXX --upgrade_mapblocks YY
```

File: src/script/lua_api/l_env.cpp [hunk 5]
```diff
@@ -1046,6 +1046,20 @@ int ModApiEnvMod::l_clear_objects(lua_State *L)
...
 
+// upgrade_mapblocks(threshold_version)
+// Upgrade all mapblocks below the threshold version
+int ModApiEnvMod::l_upgrade_mapblocks(lua_State *L)
+{
+	GET_ENV_PTR;
+
+	// Range is sanity-checked in ServerEnvironment::upgradeMapBlocks
+	int version = luaL_checkint(L, 1);
+	env->upgradeMapBlocks(version);
```
Comment (nerzhul) [inline src/script/lua_api/l_env.cpp:1057]: as i see it's done synchronously and can take ages.

I suggest to spawn a thread and do it asynchronously for non loaded blocks.

Also i suggest to implement a command line switch like we have to migrate db in order to upgrade db synchronously at startup, or a minetest.conf flag, as you like. It's clearly better than a command line from the in game admin, which will just break end user experience if playing (except if you succeed to implement it asynchronously)

**Review by nerzhul — CHANGES_REQUESTED** [review]

**Comment by sfan5** [comment]
Superseded by `--recompress`.

**Pull request closed without merging**